lib/hyperswitch.js

```diff
@@ -365,22 +331,35 @@ HyperSwitch.prototype._request = function(req, options) {
         // Prepare to call the handler with a child HyperSwitch instance
         var childHyperSwitch = this.makeChild(childReq, options);
 
-        // Call the handler with P.try to catch synchronous exceptions.
-        var reqHandlerPromise;
-        if (handler.validator) {
-            reqHandlerPromise = P.try(function() {
-                return handler.validator.validate(childReq);
-            })
-            .then(function() {
-                return handler(childHyperSwitch, childReq);
-            });
-        } else {
-            reqHandlerPromise = P.try(handler, [childHyperSwitch, childReq]);
-        }
+        var specInfo = {
+            path: match.value.path,
+            spec: handler.spec
+        };
+        var filterIdx = 0;
+        var reqHandler = function handlerWrapper(req, hyper) {
+            if (filterIdx < match.filters.length) {
+                var filter = match.filters[filterIdx];
+                filterIdx++;
+
+                if (filter.method
+                    && filter.method !== req.method
+                    && !(filter.method === 'get' && req.method === 'head')) {
+                    return handlerWrapper(req, hyper);
+                }
+
+                return filter.filter(req, hyper, handlerWrapper, filter.options, specInfo);
+            } else {
+                return P.method(handler)(req, hyper);
+            }
+        };
 
-        if (self.metrics) {
-            reqHandlerPromise = childHyperSwitch._wrapInMetrics(reqHandlerPromise, match, req);
-        }
+        // This is a hack. Pure P.try get's executed on this tick, but we wanna
+        // wrap it in metrics and access checks and start execution only afterwards.
+        // This will go away when filters are completed.
+        var reqHandlerPromise = P.resolve()
```

Comment:
@gwicke Because we need the actual call to happen on the next tick. The problem is that after running this code we synchronously modify the `hyper` state for security checks. So the actual handler should be run on the next tick. As I state in the comment this ugly hack will go away when we make security use filters in the next iteration.